beacon-chain/db/filesystem/pruner.go

```diff
@@ -0,0 +1,250 @@
+	// Safely compute the first slot in the epoch for the latest slot
+	latest = latest - latest%params.BeaconConfig().SlotsPerEpoch
+	if latest < offset {
+		return 0
+	}
+	return latest - offset
+}
+
+// Prune prunes blobs in the base directory based on the retention epoch.
+// It deletes blobs older than currentEpoch - (retentionEpochs+bufferEpochs).
+// This is so that we keep a slight buffer and blobs are deleted after n+2 epochs.
+func (p *blobPruner) prune(pruneBefore primitives.Slot) error {
+	p.Lock()
+	defer p.Unlock()
+	log.Debug("Pruning old blobs")
+	start := time.Now()
+	totalPruned, totalErr := 0, 0
+	defer func() {
+		log.WithFields(log.Fields{
+			"prunedUpToEpoch": slots.ToEpoch(pruneBefore),
+			"duration":        time.Since(start).String(),
+			"blobsPruned":     totalPruned,
+		}).Debug("Pruned old blobs")
+		blobsPrunedCounter.Add(float64(totalPruned))
+	}()
+
+	dirs, err := listDir(p.fs, ".", filterRoot)
+	if err != nil {
+		return errors.Wrap(err, "unable to list root blobs directory")
+	}
+	for _, dir := range dirs {
+		pruned, err := p.tryPruneDir(dir, pruneBefore)
+		if err != nil {
+			totalErr += 1
+			log.WithError(err).WithField("directory", dir).Error("unable to prune directory")
+		}
+		totalPruned += pruned
+	}
+
+	if totalErr > 0 {
+		return errors.Wrapf(errPruningFailures, "pruning failed for %d root directories", totalErr)
+	}
+	return nil
+}
+
+func shouldPrune(slot, pruneBefore primitives.Slot) bool {
+	return slot >= pruneBefore
+}
+
+func (p *blobPruner) tryPruneDir(dir string, pruneBefore primitives.Slot) (int, error) {
+	root := rootFromDir(dir)
+	slot, slotCached := p.slotMap.slot(root)
+	// Return early if the slot is cached and doesn't need pruning.
+	if slotCached && shouldPrune(slot, pruneBefore) {
+		return 0, nil
+	}
+
+	// At some point we'll need to list the directory to delete the files. Doing it here and
+	// deleting files one-by-one allows us to report the number of files pruned. Otherwise,
+	// with RemoveAll we'd have to do extra syscalls, and wouldn't be able to accurately report
+	// number pruned if there is a partial failure.
+	files, err := listDir(p.fs, dir, filterSsz)
+	if err != nil {
+		return 0, errors.Wrapf(err, "failed to list blobs in directory %s", dir)
+	}
+	if len(files) == 0 {
+		log.WithField("dir", dir).Warn("pruner ignoring directory with no blob files")
+		return 0, nil
+	}
+	if !slotCached {
+		slot, err = slotFromFile(path.Join(dir, files[0]), p.fs)
+		if err != nil {
+			return 0, errors.Wrapf(err, "slot could not be read from blob file %s", files[0])
+		}
+		log.WithField("path", path.Join(dir, files[0])).WithField("slot", slot).Debug("read slot from file")
+		p.slotMap.ensure(root, slot)
+		if slot >= pruneBefore {
+			return 0, nil
+		}
+	}
+	removed := 0
+	// Removing the files one by one because it is much cheaper in terms of syscalls.
+	for _, fname := range files {
+		if err := p.fs.Remove(path.Join(dir, fname)); err != nil {
```

Comment:
I'm about to switch this to just use RemoveAll because it's easier to deal with dangling `.part` files. Otherwise prunes could get wedged if a part file is orphaned by a restart or other failure.